src/coreclr/jit/optimizer.cpp

```diff
@@ -9335,6 +9337,269 @@ bool OptBoolsDsc::optOptimizeBoolsCondBlock()
+//     Theoretically there is no maximum limit on the size of the generated chain. Therefore cost
+//     checking is used to limit the maximum number of conditions that can be chained together.
+//
+bool OptBoolsDsc::optOptimizeCompareChainCondBlock()
+{
+    assert((m_b1 != nullptr) && (m_b2 != nullptr) && (m_b3 == nullptr));
+    m_t3 = nullptr;
+
+    bool foundEndOfOrConditions = false;
+    if ((m_b1->bbNext == m_b2) && (m_b1->bbJumpDest == m_b2->bbNext))
+    {
+        // Found the end of two (or more) conditions being ORed together.
+        // The final condition has been inverted.
+        foundEndOfOrConditions = true;
+    }
+    else if ((m_b1->bbNext == m_b2) && (m_b1->bbJumpDest == m_b2->bbJumpDest))
+    {
+        // Found two conditions connected together.
+    }
+    else
+    {
+        return false;
+    }
+
+    Statement* const s1 = optOptimizeBoolsChkBlkCond();
+    if (s1 == nullptr)
+    {
+        return false;
+    }
+    Statement* s2 = m_b2->firstStmt();
+
+    assert(m_testInfo1.testTree->OperIs(GT_JTRUE));
+    GenTree* cond1 = m_testInfo1.testTree->gtGetOp1();
+    assert(m_testInfo2.testTree->OperIs(GT_JTRUE));
+    GenTree* cond2 = m_testInfo2.testTree->gtGetOp1();
+
+    // Ensure both conditions are suitable.
+    if (!cond1->OperIsCompare() || !cond2->OperIsCompare())
+    {
+        return false;
+    }
+
+    // Ensure there are no additional side effects.
+    if ((cond1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+        (cond2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+    {
+        return false;
+    }
+
+    // Integer compares only for now (until support for Arm64 fccmp instruction is added)
+    if (varTypeIsFloating(cond1->gtGetOp1()) || varTypeIsFloating(cond2->gtGetOp1()))
+    {
+        return false;
+    }
+
+    // Check for previously optimized compare chains.
+    bool op1IsTestCond;
+    bool op2IsTestCond;
+    bool op1IsCondChain = FindCompareChain(cond1, &op1IsTestCond);
+    bool op2IsCondChain = FindCompareChain(cond2, &op2IsTestCond);
+
+    // Avoid cases where optimizations in lowering will produce better code than optimizing here.
+    if (op1IsTestCond || op2IsTestCond)
+    {
+        return false;
+    }
+
+    // Combining conditions means that all conditions are always fully evaluated.
+    // Put a limit on the max size that can be combined.
+    if (!m_comp->compStressCompile(Compiler::STRESS_OPT_BOOLS_COMPARE_CHAIN_COST, 25))
+    {
+        int op1Cost    = cond1->GetCostEx();
+        int op2Cost    = cond2->GetCostEx();
+        int maxOp1Cost = op1IsCondChain ? 35 : 7;
+        int maxOp2Cost = op2IsCondChain ? 35 : 7;
+
+        // Cost to allow for chain size of three.
+        if (op1Cost > maxOp1Cost || op2Cost > maxOp2Cost)
+        {
+            JITDUMP("Skipping CompareChainCond that will evaluate conditions unconditionally at costs %d,%d\n", op1Cost,
+                    op2Cost);
+            return false;
+        }
```

Comment:
Did you end up adding some handling for potentially expensive trees somewhere else?

Reply:
It's odd to me that it's possible to "defeat" the check:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static int Foo(int a, int b)
{
    if (a > 10 && (((a * a * a * a * a > 3) & (a == 20)) != false))
    {
        return 42;
    }

    return 100;
}
```
this produces:
```asm
G_M63574_IG01:              ;; offset=0000H
        A9BF7BFD          stp     fp, lr, [sp, #-0x10]!
        910003FD          mov     fp, sp
                                                ;; size=8 bbWeight=1 PerfScore 1.50
G_M63574_IG02:              ;; offset=0008H
        1B007C01          mul     w1, w0, w0
        1B007C21          mul     w1, w1, w0
        1B007C21          mul     w1, w1, w0
        1B007C21          mul     w1, w1, w0
        52800C82          mov     w2, #100
        52800543          mov     w3, #42
        71000C3F          cmp     w1, #3
        7A54C800          ccmp    w0, #20, 0, gt
        7A4A080E          ccmp    w0, #10, nzc, eq
        1A83D040          csel    w0, w2, w3, le
                                                ;; size=40 bbWeight=1 PerfScore 11.00
G_M63574_IG03:              ;; offset=0030H
        A8C17BFD          ldp     fp, lr, [sp], #0x10
        D65F03C0          ret     lr
                                                ;; size=8 bbWeight=1 PerfScore 2.00
```

Of course it's a very constructed example but I would be afraid that there are more natural looking examples.
I suppose to fix this you would need the recursive walk you wanted to avoid. I don't think that's expensive -- the code below that call `gtSetEvalOrder` and `fgSetStmtSeq` both end up doing these kinds of full tree walks anyway -- but it's not as elegant. It would make me feel a bit more at ease, however, to find and limit the cost of each "leaf" in the compare chain.

Reply:
I think we can leave it as is, it's probably fine in practice. Will leave it up to you.

Reply:
The cost of `(((a * a * a * a * a > 3) & (a == 20))` is 35, which is on the limit for combing with the `a > 10`

Whereas the cost after combining, for example, `op1 > 3 && op2 == 10 && op3 >= 12` is 32. (It's fairly high because of the two lots of EQ(AND,0)).

Looking at this again, a limit of 35 is probably too high as it's allowing 4 items to chain together. I should probably drop it to 31 to prevent `op1 > 3 && op2 == 10 && op3 >= 12`  combing with anything else.